File: crates/krishiv-python/examples/stream_transaction_count.py
#!/usr/bin/env python3
"""Real-time transaction count using a Tumbling Event-Time Window in Python."""

import pyarrow as pa
import krishiv as ks

def main():
    # 1. Build an embedded in-process session
    session = ks.Session.embedded()

    # 2. Prepare streaming mock transaction batches (timestamp, user_id)
    schema = pa.schema([
        ("timestamp", pa.int64()),
        ("user_id", pa.string()),
    ])
    batch = pa.RecordBatch.from_pydict({
        "timestamp": [1000, 2000, 61000, 62000],
        "user_id": ["Alice", "Bob", "Alice", "Alice"],
    }, schema=schema)

    # 3. Register as a bounded memory stream (watermark_column="timestamp", max_lateness_ms=5000)
    stream = session.memory_stream(
        "transactions",
        [ks.Batch(batch)],
        watermark_column="timestamp",
        max_lateness_ms=5000
    )

    # 4. Declare event-time windowing via the fluent Python API
    windowed = (
        stream
        .key_by("user_id")
        .tumbling_window(60) # 60 seconds tumbling window size (1 minute)
    )

    # 5. Execute in-process and collect output stream batches
    results = windowed.collect()
    
    # 6. Convert PyBatches to pyarrow Table and print formatted output
    pa_batches = [pa.record_batch(b.to_arrow()) for b in results]
    table = pa.Table.from_batches(pa_batches)
    print(table)

if __name__ == "__main__":
    main()

File: crates/krishiv/examples/stream_transaction_count.rs
//! Real-time transaction count using a Tumbling Event-Time Window.
//! Tested locally using in-memory streams without requiring a Kafka broker.
//! Run with: `cargo run -p krishiv --example stream_transaction_count`

#![forbid(unsafe_code)]

use std::error::Error;
use std::sync::Arc;

use arrow::array::{Int64Array, StringArray};
use krishiv::{
    DataType, ExecutionMode, Field, QueryResult, RecordBatch, Schema, Session, StreamBatch,
    WatermarkSpec,
};

fn main() -> Result<(), Box<dyn Error>> {
    // 1. Build an embedded in-process session
    let session = Session::builder()
        .with_execution_mode(ExecutionMode::Embedded)
        .build()?;

    // 2. Prepare streaming mock transaction batches (timestamp, user_id)
    let schema = Arc::new(Schema::new(vec![
        Field::new("timestamp", DataType::Int64, false),
        Field::new("user_id", DataType::Utf8, false),
    ]));
    let batch = RecordBatch::try_new(
        schema.clone(),
        vec![
            Arc::new(Int64Array::from(vec![1000, 2000, 61000, 62000])),
            Arc::new(StringArray::from(vec!["Alice", "Bob", "Alice", "Alice"])),
        ],
    )?;

    // 3. Register as a bounded memory stream (sequence=0)
    let stream_batch = StreamBatch::new(0, batch);
    let stream = session
        .memory_stream("transactions", vec![stream_batch])
        .unwrap();

    // 4. Declare event-time windowing via the fluent Rust API
    let windowed = stream
        .key_by("user_id")
        .with_event_time("timestamp")
        .watermark(WatermarkSpec::fixed_lag_ms(5000)) // 5s allowed lateness
        .tumbling_window(60000); // 1-minute tumbling window size

    // 5. Execute in-process and collect output stream batches
    let results = windowed.collect()?;

    // Extract record batches and print formatted query result
    let batches = results.into_iter().map(|b| b.batch().clone()).collect();
    println!("{}", QueryResult::new(batches).pretty()?);

    Ok(())
}
